File: tests/coretests/Stocks/Services/SingleBarPriceAnalysisTests.cs
using System.Collections.Generic;
using core.fs.Services.Analysis;
using core.Shared;
using coretests.testdata;
using Xunit;

namespace coretests.Stocks.Services
{
    public static class SingleBarPriceAnalysisFunctions
    {
        public static List<AnalysisOutcome> Generate(Ticker ticker) =>
            SingleBarPriceAnalysis.run(TestDataGenerator.PriceBars(ticker));
    }
    
    // NOTE: this tests single bar analysis with a feed of prices that do not change
    // to make sure analysis still runs and does not breakdown with various
    // exceptions related to stddev and other stats being zero
    public class SingleBarPriceAnalysisTests_FeedWithPricesNotChanging
    {
        private readonly List<AnalysisOutcome> _outcomes = SingleBarPriceAnalysisFunctions.Generate(new Ticker("SWCH"));

        [Fact]
        public void OutcomesMatch() => Assert.NotEmpty(_outcomes);
    }

    public class SingleBarPriceAnalysisTests_NewHighWorks
    {
        [Fact]
        public void Run_WithNewHigh_IncludesNewHighOutcome() =>
            Assert.Contains(
                SingleBarPriceAnalysisFunctions.Generate(new Ticker("SHEL")),
                o => o.Key == SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewHigh && o.Value == 1
            );

        [Fact]
        public void Run_WithoutNewHigh_DoesNotIncludeNewHigh() =>
            Assert.Contains(
                SingleBarPriceAnalysisFunctions.Generate(TestDataGenerator.NET),
                o => o.Key == SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewHigh && o.Value == 0
            );
    }

    public class SingleBarPriceAnalysisTests
    {
        private readonly List<AnalysisOutcome> _outcomes = SingleBarPriceAnalysisFunctions.Generate(TestDataGenerator.NET);
        
        [Fact]
        public void OutcomesMatch() => Assert.NotEmpty(_outcomes);
        
        [Fact]
        public void Volume() =>
            Assert.Equal(6933219m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.Volume).Value);
        
        [Fact]
        public void RelativeVolume() =>
            Assert.Equal(1.34m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.RelativeVolume).Value);
        
        [Fact]
        public void Open() =>
            Assert.Equal(43.95m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.Open).Value);
        
        [Fact]
        public void Close() =>
            Assert.Equal(49.14m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.Close).Value);
        
        [Fact]
        public void ClosingRange() =>
            Assert.Equal(1m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.ClosingRange).Value);
        
        [Fact]
        public void PercentChange() =>
            Assert.Equal(0.10m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.PercentChange).Value, 2);
        
        [Fact]
        public void SigmaRatio() =>
            Assert.Equal(1.62m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.SigmaRatio).Value);
        
        [Fact]
        public void GapPercentage() =>
            Assert.Equal(0m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.GapPercentage).Value);
        
        [Fact]
        public void NewLow() =>
            Assert.Equal(0, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewLow).Value);
        
        [Fact]
        public void NewHigh() =>
            Assert.Equal(0, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewHigh).Value);
        
        [Fact]
        public void SMA20Above50Days() =>
            Assert.Equal(47, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.SMA20Above50Days).Value);
        
        [Fact]
        public void PriceAbove20SMA() =>
            Assert.Equal(0.0396m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.PriceAbove20SMA).Value, 3);
        
        [Fact]
        public void TrueRange() =>
            Assert.Equal(6m, _outcomes.FirstOutcome(SingleBarPriceAnalysis.SingleBarOutcomeKeys.TrueRange).Value, 2);
    }
}

File: tests/coretests.fs/Stocks/Services/SingleBarPriceAnalysisTests.fs
module coretests.fs.Stocks.Services.SingleBarPriceAnalysisTests

open Xunit
open core.Shared
open core.fs.Services.Analysis
open testutils
open FsUnit

let generate ticker =
    Ticker(ticker) |> TestDataGenerator.PriceBars |> SingleBarPriceAnalysis.run

// NOTE: this tests single bar analysis with a feed of prices that do not change
// to make sure analysis still runs and does not breakdown with various
// exceptions related to stddev and other stats being zero
[<Fact>]
let ``SingleBarPriceAnalysisWithPricesNotChangingTests`` () =
    "SWCH"
    |> generate
    |> should not' Empty

[<Fact>]
let ``Detects new high`` () =
    "SHEL"
    |> generate
    |> MultipleBarPriceAnalysisTests.firstOutcomeGeneric SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewHigh
    |> should equal 1m

[<Fact>]
let ``When no new high, does not include new high`` () =
    "NET"
    |> generate
    |> MultipleBarPriceAnalysisTests.firstOutcomeGeneric SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewHigh
    |> should equal 0m

let outcomes = "NET" |> generate

let firstOutcome key =
    MultipleBarPriceAnalysisTests.firstOutcomeGeneric key outcomes

[<Fact>]
let ``NET outcomes are not empty``() =
    outcomes |> should not' Empty
    
[<Fact>]
let ``Volume is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.Volume |> firstOutcome |> should equal 6933219m
    
[<Fact>]
let ``RelativeVolume is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.RelativeVolume |> firstOutcome |> should equal 1.34m
    
[<Fact>]
let ``Open is correct``() =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.Open |> firstOutcome |> should equal 43.95m
    
[<Fact>]
let ``Close is correct``() =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.Close |> firstOutcome |> should equal 49.14m
    
[<Fact>]
let ``Closing range is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.ClosingRange |> firstOutcome |> should equal 1m
    
[<Fact>]
let ``Percent change is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.PercentChange |> firstOutcome |> MultipleBarPriceAnalysisTests.rounded 2 |> should equal 0.10m
    
[<Fact>]
let ``SigmaRatio is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.SigmaRatio |> firstOutcome |> should equal 1.62m
    
[<Fact>]
let ``GapPercentage is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.GapPercentage |> firstOutcome |> should equal 0m
    
[<Fact>]
let ``NewLow is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewLow |> firstOutcome |> should equal 0m
    
[<Fact>]
let ``NewHigh is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.NewHigh |> firstOutcome |> should equal 0m
    
[<Fact>]
let ``SMA20Above50Bars is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.EMA20AboveSMA50Bars |> firstOutcome |> should equal -48m
    
[<Fact>]
let ``SMA50Above200Bars is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.SMA50AboveSMA200Bars |> firstOutcome |> should equal -205m
    
[<Fact>]
let ``PriceAbove200SMA is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.PriceAboveSMA200 |> firstOutcome |> MultipleBarPriceAnalysisTests.rounded 4 |> should equal -0.2918m
    
[<Fact>]
let ``TrueRange is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.TrueRange |> firstOutcome |> should equal 6m
    
[<Fact>]
let ``DollarChange is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.DollarChange |> firstOutcome |> should equal 4.53m
    
[<Fact>]
let ``PercentDifferenceFromLow is correct`` () =
    SingleBarPriceAnalysis.SingleBarOutcomeKeys.DollarChangeVsATR |> firstOutcome |> MultipleBarPriceAnalysisTests.rounded 2 |> should equal 1.14m

